### PR TITLE
[core] Rework the field lookups in base classes to avoid iteration.

### PR DESCRIPTION
**Pull request by vgvassilev**
The past approach was slow but also does not allow us to migrate to llvm13 as
the CXXBasePath cannot take the lookup iterator anymore.

See llvm/llvm-project@0cb7e7ca0c864e052bf49978f3bcd667c9e16930

### CODE REVIEW

**Comment by smuzaffar** [comment]
https://github.com/cms-sw/root/pull/166 should test this

File: core/clingutils/res/TClingUtils.h [hunk 1]
```diff
@@ -354,7 +354,7 @@ clang::QualType AddDefaultParameters(clang::QualType instanceType,
                                      const TNormalizedCtxt &normCtxt);
 
 //______________________________________________________________________________
-llvm::StringRef DataMemberInfo__ValidArrayIndex(const clang::DeclaratorDecl &m, int *errnum = 0, llvm::StringRef  *errstr = 0);
+llvm::StringRef DataMemberInfo__ValidArrayIndex(const cling::Interpreter& interp, const clang::DeclaratorDecl &m, int *errnum = 0, llvm::StringRef  *errstr = 0);
```
Comment (Axel-Naumann) [inline core/clingutils/res/TClingUtils.h:357]:
```suggestion
llvm::StringRef DataMemberInfo__ValidArrayIndex(const cling::Interpreter& interp, const clang::DeclaratorDecl &m, int *errnum = nullptr, llvm::StringRef *errstr = nullptr);
```

File: core/clingutils/src/TClingUtils.cxx [hunk 2]
```diff
@@ -204,75 +205,28 @@ static bool IsFieldDeclInt(const clang::FieldDecl *field)
...
-   }
-   return nullptr;
+   clang::DeclarationName DName = &SemaR.Context.Idents.get(what);
+   clang::LookupResult R(SemaR, DName, clang::SourceLocation(),
+                         clang::Sema::LookupOrdinaryName,
+                         clang::Sema::ForExternalRedeclaration);
+   SemaR.LookupInSuper(R, &const_cast<clang::CXXRecordDecl&>(cl));
+   if (R.empty())
+      return nullptr;
+   return llvm::dyn_cast<const clang::FieldDecl>(R.getFoundDecl());
```
Comment (Axel-Naumann) [inline core/clingutils/src/TClingUtils.cxx:229]: Is this guaranteed to be unique?

File: core/clingutils/src/TClingUtils.cxx [hunk 3]
```diff
@@ -3130,7 +3084,7 @@ clang::QualType ROOT::TMetaUtils::AddDefaultParameters(clang::QualType instanceT
 /// If errstr is not null, *errstr is updated with the address of a static
 ///   string containing the part of the index with is invalid.
 
-llvm::StringRef ROOT::TMetaUtils::DataMemberInfo__ValidArrayIndex(const clang::DeclaratorDecl &m, int *errnum, llvm::StringRef *errstr)
+llvm::StringRef ROOT::TMetaUtils::DataMemberInfo__ValidArrayIndex(const cling::Interpreter& interp, const clang::DeclaratorDecl &m, int *errnum, llvm::StringRef *errstr)
```
Comment (Axel-Naumann) [inline core/clingutils/src/TClingUtils.cxx:3087]: Why not
```suggestion
llvm::StringRef ROOT::TMetaUtils::DataMemberInfo__ValidArrayIndex(cling::Interpreter& interp, const clang::DeclaratorDecl &m, int *errnum, llvm::StringRef *errstr)
```
?

Comment (Axel-Naumann) [inline core/clingutils/src/TClingUtils.cxx:3087]:
```suggestion
llvm::StringRef ROOT::TMetaUtils::DataMemberInfo__ValidArrayIndex(const cling::Interpreter &interp, const clang::DeclaratorDecl &m, int *errnum, llvm::StringRef *errstr)
```

File: core/dictgen/src/rootcling_impl.cxx [hunk 4]
```diff
@@ -1334,12 +1334,12 @@ void WriteNamespaceInit(const clang::NamespaceDecl *cl,
 /// array data member.
 /// In case of error, or if the size is not specified, GrabIndex returns 0.
 
-llvm::StringRef GrabIndex(const clang::FieldDecl &member, int printError)
+llvm::StringRef GrabIndex(const cling::Interpreter& interp, const clang::FieldDecl &member, int printError)
```
Comment (Axel-Naumann) [inline core/dictgen/src/rootcling_impl.cxx:1337]:
```suggestion
llvm::StringRef GrabIndex(cling::Interpreter &interp, const clang::FieldDecl &member, int printError)
```

File: core/metacling/src/TClingDataMemberInfo.cxx [hunk 5]
```diff
@@ -669,7 +669,7 @@ llvm::StringRef TClingDataMemberInfo::ValidArrayIndex() const
       return llvm::StringRef();
    }
    const clang::DeclaratorDecl *FD = llvm::dyn_cast<clang::DeclaratorDecl>(GetTargetValueDecl());
-   if (FD) return ROOT::TMetaUtils::DataMemberInfo__ValidArrayIndex(*FD);
+   if (FD) return ROOT::TMetaUtils::DataMemberInfo__ValidArrayIndex(*fInterp, *FD);
    else return llvm::StringRef();
```
Comment (Axel-Naumann) [inline core/metacling/src/TClingDataMemberInfo.cxx:673]:
```suggestion
   if (FD)
      return ROOT::TMetaUtils::DataMemberInfo__ValidArrayIndex(*fInterp, *FD);
   return {};
```

**Review by Axel-Naumann — APPROVED** [review]
Congrats, nice!!!

Some tiny comments, but if Jenkins + CMS are happy...

File: core/clingutils/src/TClingUtils.cxx (hunk 2 above)
Comment (vgvassilev) [reply core/clingutils/src/TClingUtils.cxx:229]: I don't know. I prefer to see a crash and react on it.

File: core/dictgen/src/rootcling_impl.cxx (hunk 4 above)
Comment (vgvassilev) [reply core/dictgen/src/rootcling_impl.cxx:1337]: We cannot remove the `const` without a const_cast in `WriteStreamer` then... Is that what we want?

Comment (vgvassilev) [reply core/dictgen/src/rootcling_impl.cxx:1337]: And then we need a `const_cast` or removing of the const in `llvm::StringRef TClingDataMemberInfo::ValidArrayIndex() const`.

Comment (Axel-Naumann) [reply core/dictgen/src/rootcling_impl.cxx:1337]: Sorry, I missed that call. Fine - whatever approach you prefer.

**Comment by smuzaffar** [comment]
@vgvassilev , cms tests passed.

**Comment by vgvassilev** [comment]
@smuzaffar, thanks a lot for doing these checks they are really helpful for us - and they remove a lot of stress during releases ;)

**Comment by vgvassilev** [comment]
The failures are unrelated.

**Pull request merged**